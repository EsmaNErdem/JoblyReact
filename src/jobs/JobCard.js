import React, { useContext, useState, useEffect } from "react";
import UserContext from "../auth/UserContext";

/**
 * Show short info about a job
 * 
 * - Rendered by JobDisplayList to display card for eacj job
 * - Utilizes job application for user and updates button 
 * JobDisplayList ==> JobCard
 * 
 */

const JobCard = ({ id, title, salary, equity, company }) => {
    const { hasApplied, applyJob } = useContext(UserContext);
    const [applied, setApplied] = useState()

    /**By using the useEffect, the applied status is only recalculated when the id or the hasAppliedToJob function changes, avoiding unnecessary recalculations on every render.  */
    useEffect(() => {
        console.debug("JobCard useEffect, id=", id)

        setApplied(hasApplied(id))
    }, [id, hasApplied])

    const handleApply = () =>{
        if(hasApplied(id)) return
        applyJob(id)
        setApplied(true)
    }

    return (
        <div className="JobCard">
            <h2>{title}</h2>
            <h3>{company}</h3>
            {salary && <p>Salary: ${addCommas(salary)}</p>}
            {equity && <p>Equity: {equity}</p>}
            <button 
                onClick={handleApply}
                disabled={applied}
            > 
                { applied ? "APPLIED" : "APPLY"}
            </button>
        </div>
    )
}

function addCommas(salary) {
    let digitsRev = "";
    const salaryStr = salary.toString();
  
    for (let i = salaryStr.length - 1; i >= 0; i--) {
      digitsRev = salaryStr[i] + digitsRev
      if (i > 0 && i % 3 === 0) digitsRev = "," + digitsRev
    }
  
    return digitsRev
}

export default JobCard;