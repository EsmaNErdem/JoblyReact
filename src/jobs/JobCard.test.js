import { render, cleanup, waitFor, act } from "@testing-library/react";
import JobCard from './JobCard';
import { UserProvider } from "../testUtils";

test("renders without crashing", () => {
    render(
    <UserProvider>
        <JobCard 
            id={1}
            title="job"
            salary={10000}
            equity={0.2} 
            company="C1" 
        />
    </UserProvider>
    );
});

test("it renders and matches with snaphot", () => {
    const { asFragment } = render(
        <UserProvider>
            <JobCard 
                id={1}
                title="job"
                salary={10000}
                equity={0.2} 
                company="C1" 
            />
        </UserProvider>
    );
    expect(asFragment()).toMatchSnapshot();
});

test("Display job data", () => {
    const { getByTestId } = render(
        <UserProvider>
            <JobCard 
                id={1}
                title="job"
                salary={10000}
                equity={0.2} 
                company="C1" 
            />
        </UserProvider>
        );
    
    expect(getByTestId("job-title")).toHaveTextContent("job");
    expect(getByTestId("company-name")).toHaveTextContent("C1");

});

