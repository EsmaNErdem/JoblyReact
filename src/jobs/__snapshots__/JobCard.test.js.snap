// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`it renders and matches with snaphot 1`] = `
<DocumentFragment>
  <div
    class="JobCard"
  >
    <h2
      data-testid="job-title"
    >
      job
    </h2>
    <h3
      data-testid="company-name"
    >
      C1
    </h3>
    <p>
      Salary: $100,00
    </p>
    <p>
      Equity: 0.2
    </p>
    <button>
      APPLY
    </button>
  </div>
</DocumentFragment>
`;
